middle_end/flambda/simplify/expr_builder.ml

```diff
@@ -163,6 +164,66 @@ let create_let uacc (bound_vars : BLB.t) defining_expr
     uacc,
     let_creation_result
 
+let create_coerced_singleton_let
+      uacc var defining_expr ~coercion_from_defining_expr_to_var
+      ~free_names_of_defining_expr ~body
+      ~cost_metrics_of_defining_expr =
+  if Coercion.is_id coercion_from_defining_expr_to_var then
+    create_let uacc (BLB.singleton var) defining_expr
+      ~free_names_of_defining_expr ~body ~cost_metrics_of_defining_expr
+  else
+    match (defining_expr : Named.t) with
+    | Simple simple ->
+      let defining_expr =
+        Simple.apply_coercion_exn simple coercion_from_defining_expr_to_var
+        |> Named.create_simple
+      in
+      create_let uacc (BLB.singleton var) defining_expr
+        ~free_names_of_defining_expr ~body ~cost_metrics_of_defining_expr
+    | Prim _ | Set_of_closures _ | Static_consts _ | Rec_info _ ->
+      let uncoerced_var =
+        let name = "uncoerced_" ^ Variable.raw_name (VB.var var) in
+        Variable.create name
+      in
+      (* Generate [let var = uncoerced_var @ <coercion>] *)
+      let (body, uacc, inner_result) as inner =
+        let defining_simple =
+          Simple.with_coercion (Simple.var uncoerced_var)
+            coercion_from_defining_expr_to_var
+        in
+        let defining_expr = defining_simple |> Named.create_simple in
+        let free_names_of_defining_expr =
+          Named.free_names defining_expr
+        in
+        let cost_metrics_of_defining_expr =
+          Code_size.simple defining_simple
+          |> Cost_metrics.from_size
+        in
+        create_let uacc (BLB.singleton var) defining_expr
+          ~free_names_of_defining_expr ~body ~cost_metrics_of_defining_expr
+      in
+      let generate_outer_binding name_mode =
+        (* Generate [let uncoerced_var = <defining_expr>] *)
+        let (_body, _uacc, outer_result) as outer =
+          let bound =
+            BLB.singleton (VB.create uncoerced_var name_mode)
+          in
+          create_let uacc bound defining_expr
+            ~free_names_of_defining_expr ~body ~cost_metrics_of_defining_expr
+        in
+        (* Can't somehow end up with one but not the other *)
+        assert (inner_result == outer_result);
```

Comment:
Let's write a proper equality check here; phys_equal is fragile (e.g. if we added non-constant immutable constructors to the type, this is no longer guaranteed to hold even though it should).